core/src/main/java/com/linecorp/armeria/client/HttpRequestSubscriber.java

```diff
@@ -122,43 +129,53 @@ public void operationComplete(ChannelFuture future) throws Exception {
                 return;
             }
 
-            fail(future.cause());
-        }
-
-        final Throwable cause = future.cause();
-        if (!(cause instanceof ClosedStreamException)) {
-            final Channel ch = future.channel();
-            Exceptions.logIfUnexpected(logger, ch, HttpSession.get(ch).protocol(), cause);
-            ch.close();
+            failAndWriteResetIfActive(future.cause());
         }
     }
 
     @Override
     public void onSubscribe(Subscription subscription) {
         assert this.subscription == null;
         this.subscription = subscription;
+        if (state == State.DONE) {
+            cancelSubscription();
+            return;
+        }
 
-        final EventLoop eventLoop = ch.eventLoop();
+        final HttpSession session = HttpSession.get(ch);
+        id = session.getAndIncrementNumRequestsSent();
+        if (!session.canSendRequest() || id >= MAX_NUM_REQUESTS_SENT) {
+            responseDecoder.disconnectWhenFinished();
+            // No need to send RST because we didn't send any packet and this will be disconnected anyway.
+            fail(new UnprocessedRequestException(ClosedSessionException.get()));
```

Comment:
How about adding a message to the exception? Could use different message for the two cases (`!session.canSendRequest()` and `id >= MAX_NUM_REQUESTS_SENT`)

Reply:
Specified the exception messages. Thanks!